src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs

```diff
@@ -0,0 +1,201 @@
+                subscriptionSection.AppendLine(DependencyUpdateEnd);
+                subscriptionSection.AppendLine();
+                UpdatePRDescriptionDueConfigFiles(committedFiles, subscriptionSection);
+
+                subscriptionSection.AppendLine();
+                subscriptionSection.AppendLine(sectionEndMarker);
+                _description.Insert(sectionStartIndex, subscriptionSection.ToString());
+
+            }
+            _description.AppendLine();
+            _startingReferenceId += changesLinks.Count;
+        }
+
+        private int RemovePRDescriptionSection(string sectionStartMarker, string sectionEndMarker)
+        {
+            string descriptionString = _description.ToString();
+            int sectionStartIndex = descriptionString.IndexOf(sectionStartMarker);
+            int sectionEndIndex = descriptionString.IndexOf(sectionEndMarker);
+
+            if (sectionStartIndex != -1 && sectionEndIndex != -1)
+            {
+                sectionEndIndex += sectionEndMarker.Length;
+                _description.Remove(sectionStartIndex, sectionEndIndex - sectionStartIndex);
+                return sectionStartIndex;
+            }
+            // if either marker is missing, just append at end and don't remove anything
+            // from the description
+            return _description.Length;
+        }
+
+        /// <summary>
+        /// Goes through the description and finds the biggest reference id. This is needed when updating an exsiting PR.
+        /// </summary>
+        /// <returns></returns>
+        private int GetStartingReferenceId()
+        {
+            //The regex is matching numbers surrounded by square brackets that have a colon and something after it.
+            //The regex captures these numbers
+            //example: [23]:sometext
+            Regex regex = new Regex("(?<=^\\[)\\d+(?=\\]:.+)", RegexOptions.Multiline);
+
+            return regex.Matches(_description.ToString()).Select(m => Int32.Parse(m.ToString())).DefaultIfEmpty(0).Max() + 1;
+        }
+
+        public static string GetChangesURI(string repoURI, string from, string to)
+        {
+            if (repoURI == null)
+            {
+                throw new ArgumentNullException(nameof(repoURI));
+            }
+            if (from == null)
+            {
+                throw new ArgumentNullException(nameof(from));
+            }
+            if (to == null)
+            {
+                throw new ArgumentNullException(nameof(to));
+            }
+
+            string fromSha = from.Length > 7 ? from.Substring(0, 7) : from;
+            string toSha = to.Length > 7 ? to.Substring(0, 7) : to;
+
+            if (repoURI.Contains("github.com"))
+            {
+                return $"{repoURI}/compare/{fromSha}...{toSha}";
+            }
+            return $"{repoURI}/branches?baseVersion=GC{fromSha}&targetVersion=GC{toSha}&_a=files";
+        }
+
+        public string GetPRDescription()
```

Comment:
Is there any reason not to make CalculatePRDescription return this instead of being void? Nothing stops a client from calling this method without calling Calculate first, only to find out they got an unexpected result.

Reply:
we'd be returning incomplete PR descriptions and overwriting them for every update other than the last one, which would be complete

Reply:
I don't think I follow. A call to calculatePRDescription gives you a full description for the PR given a new set of updates and a new build, doesn't it? My expectation when using this class would be something like this:

```c#
string newDescription = prDescriptionBuilder.CalculatePRDescription(params...)
```

By adding this method, you run the real risk of making mistakes like:

```c#
var prDescriptionBuilder = new PullRequestDescriptionBuilder(loggerFactory, string.empty)
string description = prDescriptionBuilder.GetPRDescription() // empty string? I thought I was getting a description!
```

and nothing is actually letting you know that you used the class wrong.

Reply:
The way it works now we call prDescriptionBuilder.CalculatePRDescription() for every update. So we only get the complete descrption after the last update. The CommitUpdatesAsync in the PullRequestActor.cs does this

Reply:
That means this class needs a very particular way of using it, which is not made clear by the public interface of the class. We can't assume that every client of this class will have the same calling pattern. 

* Nothing stops me right now from providing an empty description when initializing
* Nothing stops me from calling GetPRDescription before the full description has been calculated

By merging the two methods, you ensure that you always get a partial description for that point in time.

Reply:
I think maybe that the CalculatePRDescription isn't a very good name for the method, as it implies that it is called only once. But this is not the case we have to call this method for every repo separately. Maybe something like AppendPRDescription() would be a better name for this method, and we could rename the GetPRDescrption to ToString() or something like that

Reply:
I like that. Maybe call it AppendBuild? Since it's appending the updates from a new build?